Version 1.0

winston opened the pull request.

winston committed:
Squashed changes for Upload branch.

choonkeat committed:
Apply blindly from @choonkeat's branch. To fix in next commit.

winston committed:
Changes to sync up to @choonkeat changes.

- remove main.go (this should go into the actual project repo)
- server.go: update `uploadResponse` with `uploadMeta` struct
- server_test.go: fixed merge errors
- store.go: change method signature to use `*bytes.Reader`, add `fileType string` and return `filePath string`
- store_test.go: change dummyStore to sync with changes in store.go
- upload.go:
  - use `&bytes.Buffer` instead of ioutil -> been reading that Buffer is better for memory?
  - get file length from stream instead of `os`
  - calls `s.Storage.Upload` with `fileType` so that it returns a full `filePath` with `region`, `bucket` and `uniqueKey` and `extension`
  - uses the uploadMeta struct
  - add `meta` function that attempts to figure out three meta properties 1) datetime 2) latlong 3) geometry
 - s3store/s3.go:
  - update method signature to use `file *bytes.Reader` and `fileType string`
  - returns a full `filePath` with `region`, `bucket` and `uniqueKey` and `extension`

winston commented:
Also, have made this repo public. Can probably shift to somewhere for all `attache` stuff if you want, once it's ready for use. 🙇

choonkeat commented on s3store/s3.go:
```diff
@@ -0,0 +1,47 @@
+// Upload fulfills attache.Store interface
+func (s Store) Upload(file *bytes.Reader, fileType string) (string, error) {
+	fileName := filename(fileType)
+	filePath := fmt.Sprintf("https://s3-%s.amazonaws.com/%s/%s", os.Getenv("AWS_REGION"), s.Bucket, fileName)
```
👍 indeed full URL will make it easier to wield on the app side.

Can consider deferring to env to allow templating of this string; avoid bucket name / region issues (i recall there is some awkwardness). So if app need to configure a specific cloudfront/human friendly domain they can setup their own without too much hassle

```
FILE_URL=https://cdn.example.com/{{ .Bucket }}/{{ .Filename }}
```

choonkeat replied:
actually, i take that back. the return string should NOT be the s3 url (or google cloud storage url, or azure storage url, ...)

- after a jpg file is uploaded to attache, it will store the file as unique s3 object, e.g. "87654321.jpg"
- attache will return the `{ "path": "...", ... }` json; your app will store the necessary value in db

so far so good, right?

But your app should never render `<img src=...>` using an s3 url. Instead your app should always point to attache 
- i.e. `<img src=http://attache.example.com/87654321.jpg>`
- or `<img src=http://attache.example.com/resize/100x/87654321.jpg>`

So, why give that s3 url for your app to store?

And your app likely want to front attache GET requests with whatever cdn / varnish / custom domain, e.g. `https://random.cloudfront.net`...

And also use different processing instructions on different landing pages, e.g. `/resize/100x`

So it is really simpler (and more correct) for your app to concern itself with the url prefix of the s3 object key `87654321.jpg`, e.g. `https://random.cloudfront.net/87654321.jpg` or `https://random.cloudfront.net/resize/100x/87654321.jpg`

So, the `Store#Upload` function should only return the "filename" (aka s3 key)

winston replied:
I had the same thought too but was thinking that if we don't store the full URL in the DB, then what's in the DB is considered incomplete? i.e. you'll need the ENV of region and bucket in the app itself to piece together the information of where the image is.
(I was looking at Uploadcare as well, and they return the full URL.)

Also, is there a the possibility that the DOWNLOAD/RESIZE portion is totally divorced as a another service? So it can accept image from anywhere?

winston replied:
Working on the changes now btw.

choonkeat replied:
> what's in the DB is considered incomplete? i.e. you'll need the ENV of region and bucket in the app itself to piece together the information of where the image is.

No -- you're still thinking the "app" has to worry about the env, region, bucket.. it doesn't. the app only points to attache.

attache is the one configured with bucket, region, etc (see `s3store.Store` struct)

winston replied:
Gotcha. Let me fix in the new PR.

dmitry-mukhin replied:
I was summoned :)

In fact Uploadcare widget returns both the URL (and users can choose to store it in the app for simplicity, and it covers 80% of use cases) and file metadata (via JS API.)

Also, knowing the URL, users can get all the [file metadata](https://uploadcare.com/documentation/rest/#file) later save/use some subset of it.

choonkeat commented on s3store/s3.go:
```diff
@@ -0,0 +1,47 @@
+
+func filename(fileType string) string {
+	// Sorts in Reverse Chrono Order
+	key := strconv.FormatInt((math.MaxInt64 - time.Now().UnixNano()), 10)
```
I know it's already using `UnixNano`... but would feel more sure footed, using some sort of UUID e.g. https://github.com/oklog/ulid

choonkeat replied:
in fact, if we can suffix with the sanitized version of the uploaded filename, it would also help operationally to eyeball and see which image is which. utterly optional though

winston replied:
Am changing it to use `ulid` now.

winston replied:
Am thinking that this method should go out to `store.go`?

choonkeat replied:
👍 while implementing `gcloudstore.Store` I did feel this can be a common function for various storage engines to use. For edge cases where this cannot work, then storage can 
write its own

> Am thinking that this method should go out to store.go?

You pick any file, but it should be a standalone, public function e.g. `attache.UniqueFilename()` not attached to the `attache.Store` interface.

winston replied:
👍  New PR.

choonkeat commented on upload.go:
```diff
@@ -0,0 +1,85 @@
+}
+
+func meta(file *bytes.Reader, fileType string) uploadMeta {
+	fileMeta := uploadMeta{DateTime: "", LatLong: "", Geometry: ""}
```
either would suffice (instead of specifying empty values for each attribute)

``` go
var fileMeta uploadMeta
// or
fileMeta :=  uploadMeta{}
```

winston replied:
🆗 !

choonkeat commented on upload.go:
```diff
@@ -0,0 +1,85 @@
+func meta(file *bytes.Reader, fileType string) uploadMeta {
+	fileMeta := uploadMeta{DateTime: "", LatLong: "", Geometry: ""}
+
+	if strings.Contains(fileType, "image") {
```
`strings.HasPrefix(fileType, "image/")` to be more precise?

winston replied:
👍

choonkeat commented on upload.go:
```diff
@@ -0,0 +1,85 @@
+		xLat, xLong, xerr := x.LatLong()
+		if xerr != nil {
+			log.Println(xerr.Error())
+		}
```
let's skip these `xerr`; no values

winston replied:
Need to understand more about when we capture and bubble up errors in the idiomatic Go.

winston replied:
Will skip the `xerr` values. In this case, I'll not return any error messages actually, rendering the next comment https://github.com/winston/attache-lambda/pull/2#discussion_r157371999 somewhat invalid?

Assumption is that the extraction of file meta data can fail horribly, because the upload could be an `image` but not a `photo` (which has EXIF) and so would fail at EXIF decode. But code wise, it should continue to try and figure out as much meta as possible (moving on to finding out dimensions) instead of failing/break and return.

choonkeat replied:
> Will skip the xerr values. In this case, I'll not return any error messages actually

👌 

> rendering the next comment #2 (comment) somewhat invalid?

not really. that comment is a recommendation that returning error is a valid pattern. 

i.e. if the underlying lib you use returns error, you should not try to handle it so deep inside; just return the wrapped error

choonkeat replied:
> it should continue to try and figure out as much meta as possible (moving on to finding out dimensions) instead of failing/break and return

totally 💯 👍

choonkeat commented on upload.go:
```diff
@@ -0,0 +1,85 @@
+	file.Seek(0, 0)
+	imageSrc, _, err := image.DecodeConfig(file)
+	if err != nil {
+		log.Println(err.Error())
```
it's awkward to handle errors so deep inside the function stack; returning them makes error handling obvious sometimes, e.g. if `handleUpload` can use as

``` go
fileMeta, err := extractFileMeta(file, fileType)
if err != nil {
    return result, errors.Wrapf(err, "meta")
}
```

choonkeat commented:
could we hook up to a CI running `make test`?

choonkeat committed:
compile-time check on implementing interface

choonkeat committed:
Implements Google Cloud Storage as backend

choonkeat committed:
Change Upload signature to include ctx as first param

having context.Context provides a mechanism for requests
to terminate early / recover resources

see https://blog.golang.org/context

choonkeat committed:
use "WithContext" variant of AWS sdk!

http://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/using-requests-with-go-sdk.html
https://github.com/aws/aws-sdk-go/pull/1132

choonkeat commented:
> could we hook up to a CI running make test?

I feel we should comment out the `4x3` scenario for now (since it is a missing feature, not a failed build) and merge early instead of nested PR / branches 🙇

winston committed:
Merge pull request #5 from winston/gcloudstore

Google Cloud Storage

winston committed:
Merge pull request #7 from winston/upload-ctx

Adds additional commits to `Upload` branch

winston committed:
Fixes according to comments.

winston committed:
Make all tests pass.

But this image hasn't been auto oriented.

winston commented:
Merging first!
